sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala

```diff
@@ -1494,6 +1512,287 @@ class FileStreamSourceSuite extends FileStreamSourceTest {
+    }
+  }
+
+  class FakeFileSystem extends FileSystem {
+    val requestsExists = new mutable.MutableList[Path]()
+    val requestsMkdirs = new mutable.MutableList[Path]()
+    val requestsRename = new mutable.MutableList[(Path, Path)]()
+
+    override def exists(f: Path): Boolean = {
+      requestsExists += f
+      true
+    }
+
+    override def mkdirs(f: Path, permission: FsPermission): Boolean = {
+      requestsMkdirs += f
+      true
+    }
+
+    override def rename(src: Path, dst: Path): Boolean = {
+      requestsRename += ((src, dst))
+      true
+    }
+
+    def clearRecords(): Unit = {
+      requestsExists.clear()
+      requestsMkdirs.clear()
+      requestsRename.clear()
+    }
+
+    override def getUri: URI = throw new NotImplementedError
+
+    override def open(f: Path, bufferSize: Int): FSDataInputStream = throw new NotImplementedError
+
+    override def create(
+        f: Path,
+        permission: FsPermission,
+        overwrite: Boolean,
+        bufferSize: Int,
+        replication: Short,
+        blockSize: Long,
+        progress: Progressable): FSDataOutputStream = throw new NotImplementedError
+
+    override def append(f: Path, bufferSize: Int, progress: Progressable): FSDataOutputStream =
+      throw new NotImplementedError
+
+    override def delete(f: Path, recursive: Boolean): Boolean = throw new NotImplementedError
+
+    override def listStatus(f: Path): Array[FileStatus] = throw new NotImplementedError
+
+    override def setWorkingDirectory(new_dir: Path): Unit = throw new NotImplementedError
+
+    override def getWorkingDirectory: Path = throw new NotImplementedError
+
+    override def getFileStatus(f: Path): FileStatus = throw new NotImplementedError
+  }
+
+  test("FileStreamSourceCleaner - archive - destinations match against source pattern") {
+
+    def assertNoMove(fs: FakeFileSystem, sourcePath: Path, expectedArchivePath: Path): Unit = {
+      assert(fs.requestsExists.isEmpty)
+      assert(fs.requestsMkdirs.isEmpty)
+      assert(fs.requestsRename.isEmpty)
+    }
+
+    def assertMoveFile(fs: FakeFileSystem, sourcePath: Path, expectedArchivePath: Path): Unit = {
+      assert(fs.requestsExists.nonEmpty)
+      assert(fs.requestsExists.head === expectedArchivePath.getParent)
+      assert(fs.requestsMkdirs.isEmpty)
+      assert(fs.requestsRename.nonEmpty)
+      assert(fs.requestsRename.head === ((sourcePath, expectedArchivePath)))
+    }
+
+    // source pattern: /hello*/h{e,f}ll?
+    // base archive directory path = /hello
+    val fakeFileSystem = new FakeFileSystem()
+
+    val sourcePatternPath = new Path("/hello*/h{e,f}ll?")
+    val baseArchiveDirPath = "/hello"
+
+    val sourceCleaner = new FileStreamSourceCleaner(fakeFileSystem, sourcePatternPath,
+      Some(baseArchiveDirPath))
+
+    // file 1: /hello/helln
+    // final destination = /hello/hello/helln
+    // parent directory of final destination matches source pattern
+    val sourcePath1 = new Path("/hello/helln")
```

Comment:
When I've added `/foo1/foo2` prefix to
* sourcePatternPath
* baseArchiveDirPath
* sourcePath1
* expectedDestPath1

then test failed. I think that's because of the magic `2` in the `archive` function.

Reply:
Thanks for reporting! Will check.

Reply:
Looks like you might miss something when adding prefix.

* sourcePatternPath: `/foo1/foo2/hello*/h{e,f}ll?`
* baseArchiveDirPath: `/foo1/foo2/hello`
* sourcePath: `/foo1/foo2/hello/helln`
* expectedDestPath: `/foo1/foo2/hello/foo1/foo2/hello/helln`

Now the new expectedDestPath doesn't match against the new source path pattern.

Reply:
You might be also noticed why magic number 2 works when taking a deep look at the result.

Reply:
Yeah, now I see and fine. I've missed the additional `hello` in `expectedDestPath `.

As I understand it better I've just tried another trick:

sourcePatternPath: `/helln/foo2/hello*/h{e,f}ll?`
baseArchiveDirPath: `/helln/foo2/hello`
sourcePath: `/helln/foo2/hello/helln`
expectedDestPath: `/helln/foo2/hello/helln/foo2/hello/helln`

assertNoMove fails.

Reply:
`Expected dest path` doesn't match the source pattern. 

`/helln/foo2/hello/helln` matches source pattern path, then `/helln/foo2/hello/helln` (if it's a file) and files in `/helln/foo2/hello/helln` (if it's a directory) could pick up as source files in FileStreamSource.

This test will show how source files are being read. It would give me confusion at the first time, and I guess you are also not aware of the behavior. After reading this I guess you can understand why magic number is 2.

https://github.com/apache/spark/blob/bd8da3799dd160771ebb3ea55b7678b644248425/sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala#L736-L778

Reply:
Hmmm, it's really not simple part but the linked test helped. Now I see and looks good. Thanks!